[WIP] Split PrepPipeline into separate methods, make final interpolation optional

a-hurst opened the pull request:
<!--Thanks for contributing-->
<!--If this is your first time, please make sure to read the contributing guideline-->
<!--https://github.com/sappelhoff/pyprep/blob/master/.github/CONTRIBUTING.md-->

# PR Description

(Eventually) closes #73. Currently, this PR:

- Adds separate `remove_line_noise` and `robust_reference` methods to `PrepPipeline`.
- Makes final bad channel interpolation optional.
- Removes a bunch of unused matplotlib-based MATLAB comparison code in the `PrepPipeline` unit tests.
- Removes support for notch filter types other than `spectrum_fit`, since you can now just filter line noise however you want and then run `PrepPipeline.robust_reference` if you want to use another kind. This vastly simplifies the demands on the `filter_kwargs` API and should make the `PrepSettings` filtering options much easier to document.

There's still definitely more cleanup to do, but I figured it'd be good to get the core of this up for review sooner than later!

# Merge Checklist

<!--To merge your PR we need to first take the following points into account.-->
<!--Please just leave this checklist untouched-->

- [ ] the PR has been reviewed and all comments are resolved
- [ ] all [CI][what-is-ci] checks pass
- [ ] (if applicable): the PR description includes the phrase `closes #<issue-number>` to [automatically close an issue][auto-close-documentation]
- [ ] (if applicable): bug fixes, new features, or [API][what-is-api] changes are documented in [whats_new.rst][whats-new-file]


[what-is-ci]: https://help.github.com/en/actions/building-and-testing-code-with-continuous-integration/about-continuous-integration
[auto-close-documentation]: https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
[what-is-api]: https://en.wikipedia.org/wiki/Application_programming_interface
[whats-new-file]: https://github.com/sappelhoff/pyprep/blob/master/docs/whats_new.rst

a-hurst committed:
Remove some unused code

a-hurst committed:
Make interpolation a separate Reference method

a-hurst committed:
Remove unused MATLAB comparison test code

a-hurst committed:
Add new separate methods for prep stages

a-hurst committed:
Make remove_line_noisy only use spectrum_fit

a-hurst commented:
Whoops, looks like one of the examples relies on an undocumented attribute I removed for the sake of RAM (`self.EEG_new`). Will address that tomorrow.

a-hurst committed:
Fix black's dict complaints

a-hurst committed:
Try fixing PREP example

sappelhoff reviewed (commented):
thanks, it's great that we don't need the matplotlib stuff in the tests anymore. Also reminds me of #15

sappelhoff commented:
It is possible that I introduced these conflicts :confounded: it's because I worked on dbe2062d3c7d50a0d87775e1111062ef854c366f before pulling `master` ... the pulled master, rebased, and force pushed. That was a bad idea - I hope nothing serious got destroyed.

a-hurst commented:
> It is possible that I introduced these conflicts 😖 it's because I worked on [dbe2062](https://github.com/sappelhoff/pyprep/commit/dbe2062d3c7d50a0d87775e1111062ef854c366f) before pulling `master` ... the pulled master, rebased, and force pushed. That was a bad idea - I hope nothing serious got destroyed.

No worries, I'll rebase before I push any more work! I've set this aside for a few days while tackling some other projects, but I'll try and get this wrapped up within the next week or so :)

a-hurst committed:
Move bad chan info to dicts, rename attributes

a-hurst committed:
Add get_raw method for easier data access

a-hurst commented:
Okay @sappelhoff, I've gotten the `PrepPipeline` API rewrite to a point I think I'm happy with (though I'd love to hear your thoughts and @yjmantilla's). Some of the attributes aren't documented yet and much of it is untested, but I figured it'd be best to handle that side of things once I was sure we were all happy with the changes.

My main changes here are:

- Renamed `EEG_before_interpolation` to `EEG_post_reference` and `reference_before_interpolation` to `robust_reference_signal`
- Added a getter method `current_reference_signal` that returns `reference_before_interpolation` if interpolation hasn't been done, or the post-interpolation average reference (previously `reference_after_interpolation`) if it has (and `None` if neither have been performed, though that could be changed).
- Moved all the loose `noisy_channels_original`, `noisy_channels_before_interpolation`, `bad_before_interpolation`, `still_noisy_channels`, etc. attributes into two dicts: `noisy_info` (which contains the full noisy channel dicts) and `bad_channels` (which just contains the bad channel names), each with the keys 'original', 'post-reference', and 'post-interpolation'. This makes things easier to document and has the advantage of a more consistent/informative naming scheme.
- Added `current_noisy_info` and `remaining_bad_channels` attributes that get the current noisy info and bad channel names for the pipeline (will retrieve post-interpolation values if interpolation was used, otherwise returns the pre-interpolation values). These allow for people to try enabling/disabling interpolation for their data without having to rewrite their code to access different attributes.
- Added a new `get_raw` method, which is a more flexible version of the current `raw` getter that allows retrieving the full `mne.io.Raw` object with the EEG data from any given stage in the pipeline.

Let me know what you think!

a-hurst committed:
Fix full PREP example

a-hurst committed:
Update names and scalings for PREP example

sappelhoff commented on examples/run_full_prep.py:
```diff
@@ -168,7 +172,7 @@
 EEG_new_matlab = sio.loadmat(fname_mat2)
 EEG_new_matlab = EEG_new_matlab["save_data"]
-EEG_new = prep.EEG_new
+EEG_new = removeTrend(prep.EEG_raw, sample_rate=prep.sfreq) * 1e6
```
@a-hurst could it be that you didn't rebase this PR or did a rebase that went wrong? Because we just remove this in #102

a-hurst replied:
Nope, I somehow missed the full prep example during #102 so I'm fixing it here. My bad!

a-hurst replied:
To clarify: beforehand, these `PrepPipeline` EEG data attributes were all in microVolts which matches EEGLAB convention, so the `* 1e6` wasn't needed to compare with MATLAB results. Now that those attributes are in MNE-standard Volts, we need to convert them to uV to compare with EEGLAB data the way we do here.

sappelhoff replied:
Thanks!

sappelhoff commented:
I didn't take a look at the diff because I am not sure what diff will stay after https://github.com/sappelhoff/pyprep/pull/99#discussion_r683235569 is resolved, but I still have concerns over these two points (I am commenting only based on your summary comment):

> returns reference_before_interpolation if interpolation hasn't been done, or the post-interpolation average reference (previously reference_after_interpolation) if it has

and

> (will retrieve post-interpolation values if interpolation was used, otherwise returns the pre-interpolation values

Could we not just return a dict in both cases, with the keys `pre_interpolation` and `post_interpolation` ... and the values for these keys being `None` or empty dicts if these values have not yet been computed? I think that'd be clearer. An alternative would be to send a log message upon the getter call that says something like: "you are getting the post_interpolation values". The minimum would be to have an attribute that can clearly tell you whether the prep object is "pre" or "post" interpolation, but we might already have something like that, not sure right now.

What do you think?

a-hurst commented:
> Could we not just return a dict in both cases, with the keys `pre_interpolation` and `post_interpolation` ... and the values for these keys being `None` or empty dicts if these values have not yet been computed? I think that'd be clearer. An alternative would be to send a log message upon the getter call that says something like: "you are getting the post_interpolation values". The minimum would be to have an attribute that can clearly tell you whether the prep object is "pre" or "post" interpolation, but we might already have something like that, not sure right now.

That could work, though I still think it'd be useful to have an attribute like `current_reference_signal` so there's a consistent way to get the final reference signal regardless if interpolation was enabled or disabled. For the last project we did with PyPREP we tried the analysis pipeline both with and without final bad channel interpolation, so I think that having an API where you can toggle interpolation on and off without needing to change any attributes you access afterwards would be handy.

Instead of attributes, maybe methods in the style of the new `get_raw` function would be clearer? For example, a `get_noisy_info` method that returns the most recent noisy info available by default (like `current_noisy_info` does now), but also has an argument that lets you specify explicitly whether you want the original, post-reference, or post-interpolation noisy info.

Keep in mind that for the `PrepPipeline` API I decided against exposing interpolation as a separate method and instead made it an optional flag to `robust_reference()`, so it should hopefully be clear to users based on the settings they chose (and the documentation) whether the "current reference signal" reflects interpolated data or not.

sappelhoff commented on pyprep/prep_pipeline.py:
```diff
@@ -165,75 +163,193 @@ def __init__(
+
+        Parameters
+        ----------
+        stage : str, optional
```
I think I'd do something like this:

```suggestion
        stage : {"unprocessed", "filtered", "post-reference", "post-interpolation"}, optional
```

a-hurst replied:
That causes the line to go beyond 88 characters, is line wrap for argument types something that's supported by Numpy docstyle?

sappelhoff commented on pyprep/prep_pipeline.py:
```diff
@@ -165,75 +163,193 @@ def __init__(
+            eeg_data = stages[stage.lower()]
+            if not eeg_data:
+                raise ValueError(
+                    "Could not retrieve {stage} data, as that stage of the pipeline "
+                    "has not yet been performed."
```
```suggestion
                    f"Could not retrieve {stage} data, as that stage of the pipeline "
                    "has not yet been performed."
```

a-hurst replied:
Whoops, nice catch!

sappelhoff reviewed (commented):
> Instead of attributes, maybe methods in the style of the new get_raw function would be clearer? For example, a get_noisy_info method that returns the most recent noisy info available by default (like current_noisy_info does now), but also has an argument that lets you specify explicitly whether you want the original, post-reference, or post-interpolation noisy info.

yes! That sounds good to me, I like the new `get_raw` :+1: But I think the default should be `"current"`, not `"None"`, to be more explicit.

I approve of the remaining changes, but I am slowly losing my grasp on the potential workflows we can have with our different classes and their methods :innocent: it's been too long that I actually worked with this code, I think.

a-hurst commented:
> yes! That sounds good to me, I like the new `get_raw` 👍 But I think the default should be `"current"`, not `"None"`, to be more explicit.

Great! I'll get to work on this, as well as proper tests and docs for everything new. In that case, for the sake of simplicity I'm going to leave the internal noisy info and reference signal attributes undocumented so that the new `get_x` methods are the one clear official way for users to get at that data.

> I approve of the remaining changes, but I am slowly losing my grasp on the potential workflows we can have with our different classes and their methods 😇 it's been too long that I actually worked with this code, I think.

Hopefully cleaning up the example scripts for 0.4 will refresh us all on that front!

sappelhoff committed:
Merge branch 'master' into split_pipeline_methods
